### ISSUE TITLE
Source containing multiple directories of same name results in EEXIST mkdir

### ISSUE THREAD

**Issue by randycasburn** [gulpjs/gulp#1544]
Reopening  #422 -- UseCase: move files to a staging location as part of a lengthy stream of activities associated with moving from dev --> staging.

Assertion: If two source dirs used in gulp.src( ) have the same name and this task is executed as part of a lengthy stream of tasks, an EEXIST error is thrown. When operated as an independent task (not in a stream), the tasks flawlessly.

As stated in #422, this task (below) will work 100% of the time - flawlessly. That is, unless it is put into a stream of multiple activities. Then things go haywire and gulp throws this error (path changed for obvious reasons):

`stream.js:74
      throw er; // Unhandled stream error in pipe.
      ^
Error: EEXIST: file already exists, mkdir '/my/full/path/to/the//.staging/fonts'
    at Error (native)
Process finished with exit code 1
`

To replicate this:

1. npm install bootstrap-css-only
2. npm install font-awesome
3. npm install weather-icons
4. Create gulp task thus: 

`
var bowerDir = 'bower-components';
var staging = './.staging/';
var fonts = [
bowerDir + 'bootstrap-css-only/fonts/**/*.*',
bowerDir + 'font-awesome/fonts/**/*.*',
bowerDir + 'weather-icons/font/**/*.*'
];
gulp.task('stage-fonts', ['clean-staged-fonts'], function () {
return gulp.src(fonts)
.pipe(gulp.dest(staging + 'fonts'));
});
gulp.task('clean-staged-fonts', function () {
del(staging + 'fonts');
});
`
5. run the 'stage-fonts' task.
6. run the task as part of a long stream to produce the error.

I suspect this is due to a race condition someplace along the way.  Thanks for the help.

**Comment by randycasburn** [comment]
( I have not confirmed) To test, it may be as easy as changing the outdir var in your test case.

Open gulp/test/dest.js
change line 11 from: var outpath = join(__dirname, './out-fixtures');
to: var outpath = join(__dirname, './out-fixtures/stuff');

Run test: testWriteDir

**Comment by contra** [comment]
Can you make a git repo that reproduces this?

**Comment by randycasburn** [comment]
Sure - this setup fails every other time :-)

npm install; bower install; gulp push-to-stage;

[https://github.com/randycasburn/plurious-bughunt-gulp](https://github.com/randycasburn/plurious-bughunt-gulp)

**Comment by contra** [comment]
@randycasburn Can you provide a bare minimum case? There's a lot going on in that repo, I'd like to have this reproduced as simply as possible. If I'm reading this correctly you should be able to do this with a gulpfile containing one task and two test files in a folder.

**Comment by randycasburn** [comment]
Will have to be tomorrow...but I’ll get to it. Can you at least confirm you see the error?

**Comment by randycasburn** [comment]
OK...pushed up a much simpler git. The error is different.

**Comment by randycasburn** [comment]
Note in the simplified version. (sorry for the mess I put up earlier!)

Run: push-to-stage

 If you comment out the 'clean' task it will work fine. When you put the 'clean' task back in the sequence it will error every other run.  The error produced here is different than originally reported.

**Comment by contra** [comment]
@randycasburn The error is because you're deleting the folder and writing to it at the same time. Tasks run in parallel by default, not one after the other.

**Comment by randycasburn** [comment]
I presumed that to be the response and considered that very answer. 

Can you please take a moment then and explain the difference between what what you see in the Git vs. your documentation: [https://github.com/gulpjs/gulp/blob/master/docs/recipes/running-tasks-in-series.md](url)   

According to the docs, haven't I sequenced - not parallelized the tasks?

Thanks.

**Comment by contra** [comment]
@randycasburn No, you haven't specified that they run in series.

The URL you provided is a 404, btw.

Series would look like this in 3.9:

```js
gulp.task('b', function(){
  // do stuff
});

gulp.task('a', ['b'], function(){
  // do stuff
});

// running `gulp a` runs b then a
```

and this in 4.0:

```js
function b(){
  // do stuff
}

function a(){
  // do stuff
}

gulp.task('whatever', gulp.series(a, b))

// running `gulp whatever` runs a then b
```

**Comment by randycasburn** [comment]
ref the link 404 - really just wanted to show I RTFM

I believe I see the error of my ways here...configuring async tasks is fragile and my incorrect application of this found the weak spot.

You can close this.

**Comment by contra** [comment]
@randycasburn Yep, the task system in 3.9 was confusing. This is fixed in 4.0, I recommend using that.
